### PR TITLE
Move SPI server logic to separate `spi-server-core` crate

### PR DESCRIPTION
**Pull request by mkeeter**
This PR extracts the core logic of a SPI peripheral owner from `stm32h7-spi-server` into a standalone crate, then uses that crate to implement `stm32h7-spi-server`.

This is a step towards letting tasks exclusively own their own SPI peripherals!

(alas, this also brings the RAM footprint of a `stm32h7-spi-server` from 2044 → 2064, necessitating a size bump in TOML files)

The vast majority of the PR is moving files from `stm32h7-spi-server` to `stm32h7-spi-server-core`, but there's one other change: several SPI peripheral functions move from taking `&mut self` to `&self`.

This is required because we want both `Spi` (a handle for Idol RPC) and `SpiCore` (a locally owned SPI peripheral) to look the same to users, i.e. implementing the `SpiServer` trait from #1077.  The `Spi` RPC handle is cheaply cloneable and its functions take `&self` (because it's just a handle), which means that `SpiCore` must match it.

This in turn requires a few things:
- `SpiCore`'s mutable state (current mux, lock state) must be externally owned.  This PR puts them in `Cell`'s; `SpiCore` then owns `&'static Cell<LockState>`.  To make this simpler, there's a helper macro that builds a `SpiCore` automatically, including static storage.
- Functions on `drv_stm32h7_spi::Spi` must take `&self` instead of `&mut self`, because otherwise we can't pass around cloned `SpiCore`'s.  This is _okay_ because there's only one actual set of peripheral registers, which use `VolatileCell` internally; higher-level sequencing is the same as before, because it's managed by the `SpiCore` itself.

The PR builds on top of #1077 and should be rebased + merged after it.

### CODE REVIEW

File: drv/stm32h7-spi-server-core/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+        }
+    }
+
+    fn ready_writey<
+        'b,
+        BufRead: BufReader<'b>,
+        Read: Into<BufRead> + 'b,
+        BufWrite: BufWriter<'b>,
+        Write: Into<BufWrite> + 'b,
+    >(
```
Comment (jgallagher) [inline drv/stm32h7-spi-server-core/src/lib.rs:276]: Could we lose the `Read` and `Write` generics (in favor of calling `.into()` at the call sites)? If we could, would there be any benefit (fewer `ready_writey` instantiations)?

File: drv/stm32h7-spi-server-core/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,767 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Server task for the STM32H7 SPI peripheral.
```
[the remark this answers is not in the document]
Comment (mkeeter) [reply drv/stm32h7-spi-server-core/src/lib.rs]: Thanks, fixed in 2d370b4d9776a2cec52a3ce99b41cfbb795b8097

File: drv/stm32h7-spi-server-core/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,767 @@
...
+            device::spi1::cfg2::COMM_A::FULLDUPLEX,
+            device::spi1::cfg2::LSBFRST_A::MSBFIRST,
+            device::spi1::cfg2::CPHA_A::FIRSTEDGE,
+            device::spi1::cfg2::CPOL_A::IDLELOW,
+            device::spi1::cfg2::SSOM_A::ASSERTED,
+        );
+
+        // Configure all devices' CS pins to be deasserted (set).
+        // We leave them in GPIO output mode from this point forward.
+        for device in CONFIG.devices {
```
[the remark this answers is not in the document]
Comment (mkeeter) [reply drv/stm32h7-spi-server-core/src/lib.rs:111]: Good question!  If you look at a current `spi-driver` task, you'll see TOML that looks like this:
```toml
[tasks.spi4_driver]
name = "drv-stm32h7-spi-server"
priority = 3
max-sizes = {flash = 16384, ram = 4096}
features = ["spi4", "h753"]
uses = ["spi4"]
start = true
interrupts = {"spi4.irq" = 1}
stacksize = 872
task-slots = ["sys"]

[tasks.spi4_driver.config.spi]
global_config = "spi4"
```

(the latter block could be written as `config.spi.global_config = "spi4"`, which I find more readable)

When we combine a SPI peripheral with an owning task, we use the same `config.spi.global_config` mechanism on **that task**, e.g. for `sprot`:
```toml
[tasks.sprot]
name = "drv-stm32h7-sprot-server"
priority = 4
max-sizes = {flash = 32768, ram = 32768}
stacksize = 16384
start = true
task-slots = ["sys"]
uses = ["spi4"]
features = ["sink_test", "use-spi-core"]
interrupts = {"spi4.irq" = 0b1}
config.spi.global_config = "spi4"
```

The `spi-core` build script looks for `config.spi.global_config` in whatever task it's building, so this ends up working the same way!

File: drv/stm32h7-spi-server-core/src/lib.rs (hunk 1 above)
Comment (mkeeter) [reply drv/stm32h7-spi-server-core/src/lib.rs:276]: Good idea, this seems to work (and reduces LoC; not sure if it also produces fewer instantiations)

Done in 03a92b035972f73c952b7f8650722995290ce3d3

**Pull request merged**